enhancements/teamsync-with-oidc.md

```diff
@@ -0,0 +1,197 @@
+
+The flowchart for the Login workflow looks as depicted below:
+![](https://github.com/quay/enhancements/assets/11522230/1fcdabf6-a052-4873-9721-1dda4fc2b81d)
+
+**Endpoints in OIDC**
+
+*Authorization endpoint*
+
+According to Oauth Authorization code flow, the authorization server sends a temporary (authorization) code to the client. This code is exchanged for a token. This is a secure flow for web applications with a backend that can store credentials securely. The authorization code flow is depicted below:
+![](https://github.com/quay/enhancements/assets/11522230/9967a80b-77ec-44c4-a5d6-6a0667a8392c)
+Picture credits: https://cloudentity.com/developers/basics/oauth-grant-types/authorization-code-flow/
+
+```
+  GET /authorize?
+    response_type=code%20id_token
+    &client_id=<client_id>
+    &redirect_uri=<redirect_uri>
+    &scope=openid profile email
+    &state=af0ifjsldkj
+```
+`state`: Opaque value used to maintain state between the request and the callback.
+
+Example response:
+```
+  HTTP/1.1 302 Found
+  Location: https://server.example.com:8020/oidcclient/redirect/client01
+    code=SplxlOBeZQQYbYS6WxSbIA
+    &state=af0ifjsldkj
+```
+
+*Token endpoint*
+
+The token endpoint is initiated by the client to obtain `ID token`, `access token` and `refresh token`. This endpoint accepts an authorization_code issued to the client by the authorization endpoint. When the authorization code is validated, appropriate tokens returned in response.
+
+```
+  POST /token
+    Content-Type: application/x-www-form-urlencoded
+    Authorization: Basic auth
+
+    grant_type=authorization_code&code=<authorization_code>
+    &redirect_uri=<redirect_uri>
+```
+
+Example response:
+```
+  HTTP/1.1 200 OK
+    Content-Type: application/json
+    {
+     "access_token": "SlAV32hkKG",
+     "token_type": "Bearer",
+     "refresh_token": "8xLOxBtZp8",
+     "expires_in": 3600,
+     "id_token": "eyJ ... zcifQ.ewo ... NzAKfQ.ggW8h ... Mzqg"
+    }
+```
+
+*UserInfo endpoint*
+
+The UserInfo endpoint is a protected resource that returns claims about a user and is authenticated through an `access_token`. The response of the request is returned as a JSON object.
+
+```
+  GET /userinfo
+    Accept: application/x-www-form-urlencoded
+    access_token=<access_token>
+```
+
+Example response:
+
+```
+  HTTP/1.1 200 OK
+    Content-Type: application/json
+   {
+    "sub"          : "bob",
+    "groupIds"     : [ "bobsdepartment","administrators" ],
+    "given_name"   : "Bob",
+    "name"         : "Bob Smith",
+    "email"        : "bob@mycompany.com",
+    "phone_number" : "+1 (604) 555-1234;ext5678"
+   }
+```
+
+### Constraints
+
+* OIDC groups are synced only when a user performs a fresh signin.
+* If a user is deleted from OIDC system, there is no way to remove user's team membership from quay.
```

Comment:
That's going to be an issue... Unless there's a way to iterate over a list of users with a specific claim, you won't have a way to remove Quay users from a team, or re-sync stale teams. It's possible in LDAP since it's essentially a just a directory that we can iterate on. While in the proposed implementation, it's using a claim, and trying to do that as part of the auth flow.

In the end, you could end up with users, whose claim could be changed, while still a member of a team it shouldn't be part of anymore. Not sure if updating on user login is sufficient (persistent sessions, ...).

Reply:
This is a know constraint. When a user is deleted from OIDC, they will not be able to login to quay. So, there is no way to remove their membership. 

Can you elaborate on - 
> you could end up with users, whose claim could be changed